File: test/modulemanager/TestCheckpoint.py
import pulsar as psr

was_called=False
H=psr.create_atom([0.0,0.0,0.0],1)
class MyPyBase(psr.TestModule):
    def __init__(self,myid):
        super(MyPyBase,self).__init__(myid)
    def run_test_(self):
        policy=psr.CacheData.CachePolicy.CheckpointLocal
        global was_called
        for data in [("Some data",2.1),("Other data",H)]:
            cache_entry=self.cache().get(data[0],False)
            if cache_entry:
                if not was_called:
                    raise psr.PulsarException("Data shouldn't be there")
                elif not cache_entry==data[1]:
                    cache_entry.print(psr.get_global_output())
                    data[1].print(psr.get_global_output())
                    raise psr.PulsarException("Data is not cached value")
            elif was_called:
                raise psr.PulsarException("Data should be here")
            self.cache().set(data[0],data[1],policy)
        was_called=True

def run_test():
    tester=psr.PyTester("Testing Checkpoint Python Bindings")
    for i in range(2):
        mychk=psr.Checkpoint(psr.BDBCheckpointIO("local"),
                             psr.BDBCheckpointIO("global"))
        mm=psr.ModuleManager()
        mm.load_lambda_module(MyPyBase,"MyPyBase","Module")
        if i==1:mychk.load_local_cache(mm)
        my_mod=mm.get_module("Module",0)
        msg="Checkpointing data" if i==0 else "Uncheckpointing data"
        tester.test_call(msg,True,my_mod.run_test)
        if i==0:mychk.save_local_cache(mm)
    tester.print_results()
    return tester.nfailed()

File: test/modulemanager/TestCheckpoint.cpp
#include <pulsar/testing/CppTester.hpp>
#include <pulsar/modulemanager/Checkpoint.hpp>
#include <pulsar/modulemanager/checkpoint_backends/BDBCheckpointIO.hpp>
#include <pulsar/modulebase/TestModule.hpp>

using namespace std;
using namespace pulsar;
/* Small module for testing checkpoint.  On the first call to run_test it puts
 * a double into the cache.  On the second call
 */
struct CXXModule:public TestModule{
    CXXModule(ID_t id):TestModule(id){}
    static bool was_called;
    void run_test_(){
        auto data=cache().get<double>("Some data",false);
        if(data)
        {
            if(!was_called)
                throw PulsarException("Data shouldn't be there");
            else if(std::fabs(*data-2.1)>1e-5)
                throw PulsarException("Data isn't right");
        }
        else if(was_called)
            throw PulsarException("Data should be here");

        cache().set("Some data",2.1,CacheData::CachePolicy::CheckpointLocal);
        was_called=true;
    }
};

bool CXXModule::was_called=false;

TEST_SIMPLE(TestCheckpoint){
    CppTester tester("Testing the Checkpoint class");
    for(size_t i=0;i<2;++i)
    {
        Checkpoint mychk(make_shared<BDBCheckpointIO>("local"),
                         make_shared<BDBCheckpointIO>("global"));
        auto pmm=make_shared<ModuleManager>();
        ModuleManager& mm=*pmm;
        ModuleInfo minfo;
        minfo.name="C++ Module";

        ModuleCreationFuncs mcf;
        mcf.add_cpp_creator<CXXModule>(minfo.name);
        mm.load_module_from_mcf(minfo,"Module",mcf);
        if(i==1)mychk.load_local_cache(mm);
        auto my_mod=mm.get_module<TestModule>("Module",0);
        tester.test_member_call("call "+to_string(i),true,
                                &TestModule::run_test,my_mod.operator->());
        if(i==0)mychk.save_local_cache(mm);
    }
    tester.print_results();
    return tester.nfailed();
}

